cpp/src/parquet/arrow/path_internal.cc

```diff
@@ -0,0 +1,759 @@
+                                      AllPresentTerminalNode, AllNullsTerminalNode>;
+  std::vector<Node> path;
+  std::shared_ptr<Array> primitive_array;
+  int16_t max_def_level = 0;
+  int16_t max_rep_level = 0;
+  bool has_dictionary;
+};
+
+/// \brief Contains logic for writing a single leaf node to parquet.
+/// This tracks the path from root to leaf.
+Status WritePath(ElementRange start_range, PathInfo* path_info,
+                 ArrowWriteContext* arrow_context,
+                 MultipathLevelBuilder::CallbackFunction writer) {
+  std::vector<ElementRange> stack(path_info->path.size());
+  MultipathLevelBuilderResult builder_result;
+  builder_result.leaf_array = path_info->primitive_array;
+
+  if (path_info->max_def_level == 0) {
+    int64_t leaf_length = builder_result.leaf_array->length();
+    builder_result.def_rep_level_count = leaf_length;
+    builder_result.post_list_visited_elements.push_back({0, leaf_length});
+    return writer(builder_result);
+  }
+  stack[0] = start_range;
+  RETURN_NOT_OK(
+      arrow_context->def_levels_buffer->Resize(/*new_size=*/0, /*shrink_to_fit*/ false));
+  PathWriteContext context(arrow_context->memory_pool, arrow_context->def_levels_buffer);
+
+  auto stack_base = &stack[0];
+  auto stack_position = stack_base;
+  IterationResult result;
+  while (stack_position >= stack_base) {
+    PathInfo::Node& node = path_info->path[stack_position - stack_base];
+    // Blocks ordered roughly in likely path usage.
+    if (holds_alternative<NullableNode>(node)) {
+      result = ::arrow::util::get<NullableNode>(node).Run(stack_position,
+                                                          stack_position + 1, &context);
+    } else if (holds_alternative<ListNode>(node)) {
+      result = ::arrow::util::get<ListNode>(node).Run(stack_position, stack_position + 1,
+                                                      &context);
+    } else if (holds_alternative<NullableTerminalNode>(node)) {
+      result =
+          ::arrow::util::get<NullableTerminalNode>(node).Run(*stack_position, &context);
+    } else if (holds_alternative<FixedSizeListNode>(node)) {
+      result = ::arrow::util::get<FixedSizeListNode>(node).Run(
+          stack_position, stack_position + 1, &context);
+    } else if (holds_alternative<AllPresentTerminalNode>(node)) {
+      result =
+          ::arrow::util::get<AllPresentTerminalNode>(node).Run(*stack_position, &context);
+    } else if (holds_alternative<AllNullsTerminalNode>(node)) {
+      result =
+          ::arrow::util::get<AllNullsTerminalNode>(node).Run(*stack_position, &context);
+    } else if (holds_alternative<LargeListNode>(node)) {
+      result = ::arrow::util::get<LargeListNode>(node).Run(stack_position,
+                                                           stack_position + 1, &context);
+    } else {
+      return Status::UnknownError("should never get here.", node.index());
+    }
```

Comment:
Maybe more readable:
```suggestion
    struct {
      IterationResult operator()(NullableNode& node) { // NOLINT google-runtime-references
        return node.Run(stack_position, stack_position + 1, &context);
      }
      
      // ...
      
      ElementRange* stack_position;
      PathWriteContext& context;
    } visitor = { stack_position, context };
    
    auto result util::visit(visitor, node);
```

Reply:
Two questions:
1.  Is visit generalized across types?
2.  Do you have a sense of which the compiler can do a better job optimizing: visitor vs if statement ordered with respect to likely occurences?  Likely premature optimization either way?

Reply:
>     1. Is visit generalized across types?
Not sure what you mean by "generalized across types". Doc for `visit`: https://en.cppreference.com/w/cpp/utility/variant/visit

>     2. Do you have a sense of which the compiler can do a better job optimizing: visitor vs if statement ordered with respect to likely occurences?  Likely premature optimization either way?
Under C++11 (where `switch` may not be used at compile time) `mpark::variant` uses the array-of-function-pointers approach to unbox the variant (see https://godbolt.org/z/iY3Ji9 lines 109-110). This incurs loading from that array and the function pointer call - slightly more overhead than the if-else chain (which will probably wind up as a jump table). My *intuition* would be that there's enough other stuff in this loop even if `node` is always a `NullableNode` that you won't take a noticable hit, but I haven't done benchmarks on this.

Reply:
@emkornfield FYI `mpark::visit` does support a return value when visiting